src/Pangraph/Internal/Graph.hs

```diff
@@ -0,0 +1,206 @@
+        (Nothing, Nothing)  -> Just (e, (Just v1, Just v2))
+
+makeEdge :: [Attribute] -> (Vertex, Vertex) -> Edge
+makeEdge = Edge Nothing
+
+makeVertex :: VertexID -> [Attribute] -> Vertex
+makeVertex = Vertex
+
+edges :: Pangraph -> [Edge]
+edges p = Map.elems $ edges' p
+
+vertices :: Pangraph -> [Vertex]
+vertices p = Map.elems $ vertices' p
+
+mapEdges :: Pangraph -> Map EdgeID Edge
+mapEdges = edges'
+
+mapVertices :: Pangraph -> Map VertexID Vertex
+mapVertices = vertices'
+
+edgeByID :: EdgeID -> Pangraph -> Maybe Edge
+edgeByID key p = Map.lookup key (mapEdges p)
+
+vertexByID :: VertexID -> Pangraph -> Maybe Vertex
+vertexByID key p = Map.lookup key (mapVertices p)
+
+-- -- Operations on Pangraph
+--
+-- addEdge :: Edge -> Pangraph -> Pangraph
+-- addEdge (Edge (Just _) _ _) _ = error $ "Edge already has an ID, perhaps you should use `updateEdge`"
+-- addEdge (Edge Nothing as a) p =
+--   Pangraph (mapVertices p) newMap (1 + nextEdge' p)
+--   where
+--     newMap :: Map Int Edge
+--     newMap = Map.insert (edgeID addEdgeID) (addEdgeID) (mapEdges p)
+--     addEdgeID :: Edge
+--     addEdgeID = (Edge (Just (nextEdge' p)) as a)
+--
+-- updateEdge :: Edge -> Pangraph -> Pangraph
+-- updateEdge e p =
+--   let newMap = Map.insert (edgeID e) e (mapEdges p)
+--   in  Pangraph (mapVertices p) newMap (nextEdge' p)
+--
+-- insertVertex :: Vertex -> Pangraph -> Pangraph
+-- insertVertex v p = Pangraph (newMap) (mapEdges p) (nextEdge' p)
+--   where
+--     newMap :: Map VertexID Vertex
+--     newMap = Map.insert (vertexID v) v (mapVertices p)
+--
+-- removeEdge :: Edge -> Pangraph -> Maybe Pangraph
+-- removeEdge e p = Map.
+--
+-- removeVertex :: Vertex -> Pangraph -> Maybe Pangraph
+
+edgeAttributes :: Edge -> [Attribute]
+edgeAttributes = edgeAttributes'
+
+vertexAttributes :: Vertex -> [Attribute]
+vertexAttributes = vertexAttributes'
+
+edgeEndpoints :: Edge -> (Vertex, Vertex)
+edgeEndpoints = endpoints'
+
+edgeID :: Edge -> EdgeID
+edgeID edge = fromMaybe
+  (error $ "Fatal: Edge missing ID, " ++ show edge)
+  (edgeID' edge)
+
+
+vertexID :: Vertex -> VertexID
+vertexID = vertexID'
+
+lookupVertexValues :: Vertex -> Key -> Maybe Value
+lookupVertexValues v k = lookup k (vertexAttributes v)
+
+lookupEdgeValues :: Edge -> Key -> Maybe Value
+lookupEdgeValues e k = lookup k (edgeAttributes e)
+
+-- Utility Operations
+
+vertexToAssocList :: [Vertex] -> [(VertexID, Vertex)]
+vertexToAssocList = map (\v -> (vertexID v, v))
+
+edgeToAssocList :: [Edge] -> [(EdgeID, Edge)]
+edgeToAssocList = map (\e -> (edgeID e, e))
+
+edgeContainsKey :: Key -> Edge -> Bool
+edgeContainsKey k e = elem k $ map fst $ edgeAttributes e
+
+vertexContainsKey :: Key -> Vertex -> Bool
+vertexContainsKey k v = elem k $ map fst $ vertexAttributes v
+
+instance Show Pangraph where
+  show p = "Pangraph " ++ show (vertices p) ++ " " ++ show (edges p)
+
+instance Show Vertex where
+  show v@(Vertex _ as) = unwords ["Vertex", show (vertexID v), show as]
```

Comment:
Usually it is assumed that `show x` produces code that can be compiled to obtain `x` back. This is not the case currently, because `Vertex` constructor is not exported. But `makeVertex` is, so I suggest to use it in the `show` method, i.e. `unwords ["makeVertex", show (vertexID v), show as]`.